### ISSUE TITLE
Filtering includes why different behaviour for index vs read?

### ISSUE THREAD

**Issue by edeis53** [cloudcreativity/laravel-json-api#304]
I can successfully filter an included resource on an index request, but cannot perform the same filter when requesting on a read request for a single record.

Adapter filter
```
    protected function filter($query, Collection $filters)
    {
        if ($variantClass = $filters->get('images')['variants']['class']) {
            $query->with(
                ['images.variants' => function ($subquery) use ($variantClass) {
                    $subquery->whereIn('class', $variantClass);
                }]
            );
        }
    }
```

Validator
```
protected $allowedFilteringParameters = ['images'];
```

Caveat, my api implementation uses binary uuid. Which I encode/decode where necessary. But not sure if it plays a factor in this case.
eg. Schema
```
    public function getId($model)
    {
        return (string) $model->resolveUuid()->fromBytes($model->id)->toString();
    }
```

Successful filtering query:
`items?filter[page][id]=41dbd12d-8e55-4f67-8298-27943ddc64b9&filter[images][variants][class][]=1&include=images.variants`
** Filtering by page.id here will return the same item.id as the read query below.

Corresponding relevant lines of SQL log, small edit with BINARYUUID for readability:
```
select * from `image_variants` where `image_variants`.`image_id` in (?) and `class` in (?) [BINARYUUID, 1]
```

Not successful filtering query:
`items/5e272480-6c0f-4721-bcc0-f6b63edbc179?filter[images][variants][class][]=1&include=images.variants`

SQL Log:
```
select * from `image_variants` where `image_variants`.`image_id` in (?) [BINARYUUID]
```

The if-statement to apply the filter is true on both queries. But for some reason it is not performing the filter?

**Comment by lindyhopchris** [comment]
Filtering on a read request is integration tested in this package - tests are passing. So I'm not sure what the problem is based on the info you've posted above.

Can you look into what's happening on these lines in your implementation?
https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/AbstractAdapter.php#L165-L167

**Comment by edeis53** [comment]
What should I be looking for on those lines?

`$parameters->getFilteringParameters()` show the correct filtering parameters.
```
       array:1 [
  "images" => array:1 [
    "variants" => array:1 [
      "class" => array:1 [
        0 => "1"
      ]
    ]
  ]
]
```
The result of `$record = $this->readWithFilters($record, $parameters);` is unchanged. `$record` has the same value as when passed in.

Then I traced that down to `readWithFilters` -> `applyFilters` -> to the `filter` function on my adapter.  Seems to reach all the relevant spots. Maybe there is something to do with load method line 170 that I need to adapt to work with uuid `$this->load($record, $parameters);`?

If I use a simple filter such as `$query->where('id', 1);` it works no problem on the read request, just can't do anything on a relation.

In the mean time, I've added a filter function into to the relation data returned in the schema, so my needs are being met. If your tests are passing for filtering relationship includes, then maybe something funny with my uuid adaptation.

**Comment by lindyhopchris** [comment]
Having re-read your original post I think the thing that is slightly funky is you're using the filters for something that's an include path.

`$query->with` is about eager loading, i.e. a JSON API include path not a filter.

Do you not mean to use `$query->whereHas()`? I.e. if it is actually a filter, surely your code should be:

```php
    protected function filter($query, Collection $filters)
    {
        if ($variantClass = $filters->get('images')['variants']['class']) {
            $query->whereHas('images.variants', function ($subquery) use ($variantClass) {
                    $subquery->whereIn('class', $variantClass);
            });
        }
    }
```

Or something along those lines because that is actually a filter?

**Comment by edeis53** [comment]
That would work well if I wished to filter the `items` based on whether or not they had the corresponding class. In my case, I wish not to filter the item, but filter the records in included relationship only.

eg. Scenario A) get all `items`, and include`image.variants` and filter `image.variants` with `class = 1`

Scenario B) get record `items` with `id = 2` and include`image.variants` and filter `image.variants` with `class = 1`

In summary: Scenario A currently works with the $query->with() example I have above, but Scenario B for some reason does not. Perhaps Scenario A's success was an unintended side effect here.

No worries. Perhaps this is not how filtering was intended to be implemented in this library as the json spec is agnostic about the filtering strategies.

**Comment by lindyhopchris** [comment]
It's possible I don't fully understand your use-case, but my tip would be that if you want to filter a relationship, you should use a relationship endpoint.

I.e. something along the lines of `GET /api/items/1/image-variants?filter[class]=foo` will give you the image-variants that belong to item 1 but only have the class foo.

**Comment by lindyhopchris** [comment]
Actually the relationship endpoint is what you want to use.

When you use `GET /api/items/1?filter[foo]=bar` you are filtering the item number 1, **not** the included resources. So you're saying: give me item number 1 if it's foo field equals bar.

When you use `GET /api/items/1/image-variants?filter[baz]=bat` you're filtering the image variants, not the item. So you're saying: me the the image variants that belong to item 1, only if their baz field equals bat.

**Comment by edeis53** [comment]
That nicely clears up the difference on the endpoints filtering intents. 

In regards to clarifying my use case, when requesting `items`, we are also using `include` to pull down 8 or more other model types in a single request that may require filtering. Ideally, we'd like to avoid having to make an additional 4-5 network requests to pull down those relations individually with their corresponding filters.

As filtering implementation spec is agnostic, I may have wrongly assumed we could dictate whatever filtering methods we wanted inside the filter method because the filter function simply just matches a series of `if` statements to trigger corresponding query builders.

eg. `GET /api/items/1?filter[foo]=bar` to filter as you've described, 

and for my use-case here's an idea for a clearer semantic: `GET /api/items/1?include=images,translations,comments&filter[includes][images][size]=3&filter[includes][translations][lang]=en_GB` which otherwise would have to be accomplished with three network requests using relationship endpoints.

As I mentioned early, I've implemented a work around to satisfy our needs and to filter the data returned on include. It's not as elegant as being able to use the existing filter api, but it works and we can grab all the includes and filter them in a single request.
Schema:
```
            'variants' => [
                self::SHOW_SELF => true,
                self::SHOW_RELATED => true,
                self::SHOW_DATA => isset($includedRelationships['variants']),
                self::DATA => function () use ($resource) {
                    return $this->filterVariants($resource->variants);
                },
            ],
```

Which basically does this, plus does validation of filter params, rules, etc.
```
if (isset(request()->query()['filter']['includes']['image']['variants']['state'])) {
....validated, etc....
$filtered = $variants->filter(function ($record) use (
                $state
            ) {
                //return only records with the correct state
                if ($record->state === $state) {
                    return true;
                }
            });
            return $filtered;
```

**Comment by lindyhopchris** [comment]
You're right, as the spec says it's implementation agnostic then it's totally fine for you to be interpreting it this way!

Actually your solution is probably the way to go about it - i.e. you need to filter within the schema, as what your filter is doing is changing the `data` member of the relationship, rather than the top-level `data` member.

The `neomerx/json-api` package that does all the encoding does not envisage this being done, which is why you don't have access to the filters within the `getRelationships` method... They probably wouldn't support this because they are extremely focused on the encoding being performant, and filtering within the schema is going to be costly if you've got 10s/100s of records that you're encoding.

Glad you found a way to do it at the mo. If this comes up from other people as well then it might need to be a feature to support... The feature is basically some sort of eager loading filtering. But think I'll leave it as edge-case for the moment as this looks complicated so wouldn't want to go down that route unless there was a lot of demand for it.

**Comment by edeis53** [comment]
Good to know! Thanks for your attention on this. Greatly helps to clarify how/why things are working the way they are.

**Comment by dimitvassilev** [comment]
Hello,
For what it's worth, I've found myself in a similar situation and would love to see this supported. 

I have a relationship 'labels' with a 'language' attribute in the related model and when including this relationship I only need the resources in a particular language. There are multiple relationships that have to be included simultaneously in addition to 'labels' and the primary data is a collection, rather than a single model, so I cannot get the included data even in one additional call to the relationship endpoint. It could be done via 'labels' as main resource, filtering on a foreign key with 'whereIn' logic, but only so long as the collection from the original request is small in size, which in my case it might not be. Another option is a pre-filtered relationship, e.g. 'label_en' for English, but this bloats the api when translations in many languages are involved and for multiple resource types, so I think I'll use your approach. Thanks for sharing it!

**Comment by edeis53** [comment]
I also do language filtering as well for the includes using the method above. To avoid API endpoint bloat, I send the details about language in the request headers and parse correspondingly in the filter. Makes for a very smooth developer experience.
